### ISSUE TITLE
404 Not found on images in sub folders

### ISSUE THREAD

**Issue by yngwi** [medusa-project/cantaloupe#123]
Hi,

I have the following problem: I tried to set up an cantaloupe instance on one of our servers and when I try to access an image I get a 404 error for files located in a sub path on the server.

For instance:

Working: http://iiif.icar-us.eu/iiif/2/1-19.jpg/info.json
Not working: http://iiif.icar-us.eu/iiif/2/img%2FAT-ADG%2F1-19.jpg/info.json

The problem seems to lie in the URL-encoded file name. The second request results in this error message: 
`Failed to resolve img%2FAT-ADG%2F1-19.jpg to /opt/tank2/images/monasterium/img%2FAT-ADG%2F1-19.jpg`

The path to the server seems to be correct apart from the encoded `/`.

The server is hosted by an Apache 2 instance. The virtual host is set up according to the Cantaloupe user manual. It includes `AllowEncodedSlashes NoDecode`. I enabled a static FilesystemResolver with `/opt/tank2/images/monasterium/` as the path prefix.

Can you help me figure out why this isn't working?

Kind regards,
Daniel

**Comment by adolski** [comment]
Hi,

Your configuration sounds right to me. It's interesting that FilesystemResolver is seeing an encoded identifier. Identifiers are supposed to be decoded immediately by the endpoints before passing through the rest of the application, and I can't see why this wouldn't be happening.

I wonder if the image server is somehow receiving a "double-encoded" URL. In that case it would be decoding `%252F` to `%2F`.

When I request an image from a subfolder through a reverse proxy (which I can do successfully), I see these two lines in my log:

```
INFO  e.i.l.c.r.AbstractResource - doInit(): handling GET http://localhost:8182/iiif/2/subfolder%2Fimage.jpg/info.json
INFO  e.i.l.c.r.FilesystemResolver - Resolved subfolder/image.jpg to /Users/alexd/Sites/iiif-test/subfolder/image.jpg
```

(Note decoded slashes in the second message.)

What happens when you access the image server directly, not through a reverse proxy?

What happens when you set `slash_substitute = HELLO` and replace `%2F` with `HELLO` in your URLs?

Are you able to build a snapshot from a clone using Maven?

**Comment by yngwi** [comment]
OK maybe I had a configuration error on my side: your comment brought me to having a look a the HTTP settings again and there was no URI Slash Substitute configured. I entered `%2F` and now everything works. I thought `%2F` was the default version? Or am I still wrong about something and it is only a coincidence that things seem to work now?

Thanks,
Daniel

**Comment by yngwi** [comment]
Well, do you think I will run into problems due to this or should I just roll with it?

**Comment by adolski** [comment]
It's probably fine as a workaround. I will add some debug-level log messages in 3.3.2 that will record the state of the identifier as it arrives and is decoded, which should make whatever is happening a little more transparent.

**Comment by yngwi** [comment]
OK thanks a lot for your help and the great software! Apart from this small hiccup installing it was easy as a breeze. I can't wait to use it a bit more!

**Comment by adolski** [comment]
Thanks, happy to help. The commit above is available on the `release/3.3` branch and will be part of 3.3.2. For each request, it will produce a log message like:

```
DEBUG e.i.l.c.r.AbstractResource - Identifier requested: subfolder%2Fimage.jpg / decoded: subfolder/image.jpg / slashes substituted: subfolder/image.jpg
```

The first one is straight from the URL, and the last one is what will actually be used. Please reopen when the time comes if this doesn't help you troubleshoot what's going on.
